builtin/logical/database/rotation.go

```diff
@@ -0,0 +1,380 @@
+	for {
+		// quit rotating credentials if shutdown has started
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+		item, err := b.popItem()
+		if err != nil {
+			if err == queue.ErrEmpty {
+				return nil
+			}
+			return err
+		}
+
+		// guard against possible nil item
+		if item == nil {
+			return nil
+		}
+
+		// validate the role still exists
+		role, err := b.StaticRole(ctx, s, item.Key)
+		if err != nil {
+			b.logger.Warn("unable load role", "role", item.Key, "error", err)
+			continue
+		}
+		if role == nil {
+			b.logger.Warn("role not found", "role", item.Key, "error", err)
+			continue
+		}
+
+		// if "now" is less than the Item priority, then this item does not need to
+		// be rotated
+		if time.Now().Unix() < item.Priority {
+			if err := b.pushItem(item); err != nil {
+				b.logger.Warn("unable to push item on to queue", "error", err)
+			}
+			// break out of the for loop
+			break
+		}
+
+		input := &setStaticAccountInput{
+			RoleName: item.Key,
+			Role:     role,
+		}
+
+		// If there is a WAL entry related to this Role, the corresponding WAL ID
+		// should be stored in the Item's Value field.
+		if walID, ok := item.Value.(string); ok {
+			walEntry := b.findStaticWAL(ctx, s, walID)
+			if walEntry != nil && walEntry.NewPassword != "" {
+				input.Password = walEntry.NewPassword
+				input.WALID = walID
+			}
+		}
+
+		resp, err := b.setStaticAccount(ctx, s, input)
+		if err != nil {
+			b.logger.Warn("unable to rotate credentials in periodic function", "error", err)
+			// Increment the priority enough so that the next call to this method
+			// likely will not attempt to rotate it, as a back-off of sorts
+			item.Priority = time.Now().Add(10 * time.Second).Unix()
+
+			// preserve the WALID if it was returned
+			if resp != nil && resp.WALID != "" {
+				item.Value = resp.WALID
+			}
+
+			if err := b.pushItem(item); err != nil {
+				b.logger.Warn("unable to push item on to queue", "error", err)
+			}
+			// go to next item
+			continue
+		}
+
+		lvr := resp.RotationTime
+		if lvr.IsZero() {
+			lvr = time.Now()
+		}
+
+		// update priority and push updated Item to the queue
+		nextRotation := lvr.Add(role.StaticAccount.RotationPeriod)
+		item.Priority = nextRotation.Unix()
+		if err := b.pushItem(item); err != nil {
+			b.logger.Warn("unable to push item on to queue", "error", err)
+		}
+	}
+	return nil
+}
+
+// findStaticWAL loads a WAL entry by ID. If found, only return the WAL if it
+// is of type staticWALKey, otherwise return nil
+func (b *databaseBackend) findStaticWAL(ctx context.Context, s logical.Storage, id string) *setCredentialsWAL {
+	wal, err := framework.GetWAL(ctx, s, id)
+	if err != nil {
+		b.Logger().Warn("error reading WAL", "id", id, "error", err)
+		return nil
+	}
+
+	if wal == nil || wal.Kind != staticWALKey {
+		return nil
+	}
+
+	data := wal.Data.(map[string]interface{})
+	walEntry := setCredentialsWAL{
+		walID:       id,
+		NewPassword: data["NewPassword"].(string),
```

Comment:
any way that we could format these as `new_password`? which is the format used for map keys (and json keys)

Reply:
I know this is being handled outside of the backend, so no worries if not - just curious if it's easy or not

Reply:
I added `json` tags to the fields in f5c7a9c74, and changed the keys to lower, camel case